### ISSUE TITLE
Add contactInformation-alike for softwareVersion or researchProductVersion

### ISSUE THREAD

**Issue by jagru20** [openMetadataInitiative/openMINDS_core#159]
In a discussion on how to map the former `help` -Attribute onto the openMINDS schema, @olinux and I have been struggling to find a proper equivalent. `help` can contain several entries, from which each can be a url or a mail address or similar. This is meant as an answer to a possible user's question "How can I do XY with this software?" assuming, that the answer is not in the documentanion. 
We considered `otherContributions` in combination with the contributionType `softwareSupport` but the problem here is that it cannot store information other than that there was a contribution and by whom, not where to find it. 
To solve this, we suggest adding an optional property to `softwareVersion `or `researchProductVersion` that stores this information in a string.

**Comment by lzehl** [comment]
Mh. I see your point. But I would like to have a more concrete example to be able to think about it and see if that might be also useful for the other research products.

You said "help" in the past linked to 
+ a URL (so a support page, forum page, etc, I suppose?)
+ email (which would be a person)

Any other use cases already?

**Comment by jagru20** [comment]
Thanks for the quick reply.

> You said "help" in the past linked to
> 
>     * a URL (so a support page, forum page, etc, I suppose?)
this is correct.

>     * email (which would be a person)

This is partly correct. Take `NEST v2.20.0` as an example: It has three entries at `help` {
1."support@humanbrainproject.eu"
1."https://www.nest-simulator.org/mailinglist/postorius/lists/users.nest-simulator.org/"
1. "https://github.com/nest/nest-simulator/issues"
 
All of these are either URLs or email, but for example, the mentioned email is not necessarily bound to a specific person. Does this bring some clarification? I will still have a look for more use cases though.

**Comment by lzehl** [comment]
@jagru20 thanks for the clarification. In case of general support emails I would use an organization instead of a person. 
That should fit form my perspective because an email is addressed to some legal person or organization I would say. 

So in your case: HBP, or do you think that this is incorrect?

I'm still thinking on how to solve this. Easiest would be to have as you suggested a "support" or "contact" property in the research product versions. Could you figure out if it is possible to define two formats ("IRI" and "email") for a string according to the JSON draft (we mostly are inspired by 7, I think)?

Maybe it is even worth thinking of defining a schema for this where also small instructions on how to get best support for specific issues could be entered as a free text...

@apdavison , @skoehnen , @olinux , @UlrikeS91 it would be good to hear also your thoughts on this issue. Thanks in advance :)

**Comment by jagru20** [comment]
> So in your case: HBP, or do you think that this is incorrect?

Yes. I would also go with using the organization instead of a person, but I'm not sure if we can also apply this to every other type of entries here. 

> Could you figure out if it is possible to define two formats ("IRI" and "email") for a string according to the JSON draft (we mostly are inspired by 7, I think)?

I think this is not possible (at least [this validator](https://www.jsonschemavalidator.net/) wont accept anything other than a single string for `format`).

**Comment by jagru20** [comment]
>Maybe it is even worth thinking of defining a schema for this where also small instructions on how to get best support for specific issues could be entered as a free text...

I think it is.  Could look like: 
```
{
    "_type": "https://openminds.ebrains.eu/core/FindHelpAt"
    "properties": {
        "mail": {
            "type": "string",
            "format": "idn-email",
            "_instruction": "Add an E-Mail address where the user can request help for this product."
        },
        "iri": {
            "type": "string",
            "format": "iri",
            "_instruction": "Add the internationalized resource identifier (IRI) to a help resource for this product."
        }
        "isProvidedBy":{
            "type": "array",
            "minItems": 1,
            "uniqueItems": true,
            "_instruction": "Add one or several persons or organisations that provide help for this product.",
            "_linkedTypes": [
                "https://openminds.ebrains.eu/core/Organization",
                "https://openminds.ebrains.eu/core/Person"
            ]
        }
    }
}
```

just as a suggestion... I would also suggest placing this under schemas/actors.

**Comment by lzehl** [comment]
@jagru20 thanks for further thinking about this. My thoughts here are the following: 

Normally the support email should work over a long (or endless) time. Maybe therefore the properties "email" + "iri" in the FindHelpAt schema is sufficient (both optional). 

In any case "isProvidedBy" will be ignored if "mail" is given in properties... I also realized that for Organizations and Persons the contactInformation is handled in protected space within the KG and eventually confidential which does not seem efficient for a support... Moreover, if a support is provided by a Persons it could still be covered by using "main contact" via "otherContribution". And if support is provided by an Organization I assume they still have a specific support email or a specific support webpage for the corresponding project which could then also directly be entered for the reduced FindHelpAt schema.

Now if it only breaks down to "email" and "iri" for the FindHelpAt schema and both of those should of course also be always visible (so they do not need to be in a protected space) I would maybe even further reduce this and do not provide another schema but actually immediately include those two properties on the research product version schema (e.g., as "supportEmail", "supportWebpage"; I would prefer having that even combined in "providedSupportChannel", but as you figured out two string formats cannot be handled by a validator, and I guess on regex pattern allowing either an email or an iri is also not possible?). 

@jagru20 what do you think?
@olinux & @apdavison what are your thoughts on this?

**Comment by UlrikeS91** [comment]
> I'm still thinking on how to solve this. Easiest would be to have as you suggested a "support" or "contact" property in the research product versions. Could you figure out if it is possible to define two formats ("IRI" and "email") for a string according to the JSON draft (we mostly are inspired by 7, I think)?

If I'm not mistaken, you are allowed to define two formats like this:
```
"properties": {
    "contact": {
      "type": "string",
      "anyOf":[
        {"format": "IRI"},
        {"format": "email"}
      ]
    }
  }
```
I also tried it in the validator @jagru20 linked ([this validator](https://www.jsonschemavalidator.net/)), and that was allowed ;)

**Comment by UlrikeS91** [comment]
> Now if it only breaks down to "email" and "iri" for the FindHelpAt schema and both of those should of course also be always visible (so they do not need to be in a protected space) I would maybe even further reduce this and do not provide another schema but actually immediately include those two properties on the research product version schema (e.g., as "supportEmail", "supportWebpage"; I would prefer having that even combined in "providedSupportChannel", but as you figured out two string formats cannot be handled by a validator, and I guess on regex pattern allowing either an email or an iri is also not possible?).
> 
> @jagru20 what do you think?
> @olinux & @apdavison what are your thoughts on this?

If I'm correct and we can define two formats, I would support having a single property (and also if not actually...). I would just call it "supportChannel" (without provided). 
I would not make this an additional schema. The custodian is still there. This can be an organization or a person, which can be contacted as well.

**Comment by lzehl** [comment]
@UlrikeS91 could you double check with @olinux if the double formats works and for simplicity reduce this in our template dialect to `"format": ["email", "iri"]`?

I agree with your suggestion for the property name. Thanks for jumping in as well :)

**Comment by lzehl** [comment]
I discussed with @olinux and we decided the following:

In general, the json-schema property "format" will be changed to a schema-template property "_format" expecting always (!) a list.
In cases with only one format this will look like this: `"_format": ["email"]`
In case of multiple possible formats this will look like this: `"_format": ["email", "iri"]`

I'll prepare a corresponding PR. This will be a PR that also needs further changes (commits) by @olinux before merging, because it requires an extension of the generator.
